language/libra-tools/writeset-transaction-generator/src/writeset_builder.rs

```diff
@@ -0,0 +1,199 @@
+        })
+        .collect()
+}
+
+fn move_module_changes(modules: &[CompiledModule]) -> MoveChanges {
+    let mut shadow_changeset = MoveChanges::new();
+    for module in modules.iter() {
+        let id = module.self_id();
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        shadow_changeset.publish_or_overwrite_module(id, module_bytes);
+    }
+    shadow_changeset
+}
+
+fn shadow_reconfiguration<S: StateView>(state_view: &S, move_changes: &mut MoveChanges) {
+    let move_vm = MoveVM::new();
+    let state_view_storage = RemoteStorage::new(state_view);
+    let mut session = move_vm.new_session(&state_view_storage);
+    session
+        .execute_function(
+            &ModuleId::new(
+                account_config::CORE_CODE_ADDRESS,
+                Identifier::new("LibraConfig").unwrap(),
+            ),
+            &Identifier::new("reconfigure_").unwrap(),
+            vec![],
+            vec![],
+            libra_root_address(),
+            &mut CostStrategy::system(&ZERO_COST_SCHEDULE, GasUnits::new(100_000_000)),
+            &NoContextLog::new(),
+        )
+        .unwrap();
+
+    let effect = session.finish().unwrap();
+    for (account, ops) in effect.resources {
+        for (struct_tag, op) in ops {
+            match op {
+                Some((layout, value)) => {
+                    let blob = value.simple_serialize(&layout).unwrap();
+                    move_changes
+                        .publish_resource(account, struct_tag, blob)
+                        .unwrap();
+                }
+                None => move_changes
+                    .unpublish_resource(account, struct_tag)
+                    .unwrap(),
+            }
+        }
+    }
+}
+
+pub fn build_changeset<S: StateView, F>(
+    state_view: &S,
+    override_move_changes: Option<MoveChanges>,
+    procedure: F,
+    to_be_published_modules: &[CompiledModule],
+) -> ChangeSet
+where
+    F: FnOnce(&mut GenesisSession<DeltaStorage<RemoteStorage<S>>>),
+{
+    let move_vm = MoveVM::new();
+    let mut move_changes = override_move_changes.unwrap_or_else(MoveChanges::new);
+    shadow_reconfiguration(state_view, &mut move_changes);
+    let mut effect = {
+        let state_view_storage = RemoteStorage::new(state_view);
+        let exec_storage = DeltaStorage::new(&state_view_storage, &move_changes);
+        let mut session = GenesisSession(move_vm.new_session(&exec_storage));
+        procedure(&mut session);
+        session
+            .0
+            .finish()
+            .map_err(|err| format_err!("Unexpected VM Error: {:?}", err))
+            .unwrap()
+    };
+
+    for module in to_be_published_modules {
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        // TODO: Check compatibility between old and new modules.
+        effect.modules.push((module.self_id(), module_bytes));
+    }
```

Comment:
I don't think the check here is sufficient, and conceptually I wouldn't think of it as "we need to check whether the new modules are compatible" but rather "when the effects are applied to the base state, will the new state be compatible", which includes
- Checking if the modules are compatible.
- Checking if the resources are in good state -- we must not publish a resource that's defined in a module that's part of the overriding changes.